"""
__authors__: Kevin Chau, Hanqing Zhao, Weixiu Dong
__date__: 2019-03-05
__description__: RNA-Seq alignment and quantification
"""

import os

configfile: "../../config.yaml"

# Setup pipeline directories
if not os.path.exists("data"):
    os.makedirs("data")
if not os.path.exists("log"):
    os.makedirs("log")

rule cutadapt_trim:
    input:
        read1 = os.path.join("../../source/", config["FASTQS"]["read1"] + ".fastq"),
        read2 = os.path.join("../../source/", config["FASTQS"]["read2"] + ".fastq")
    output:
        trim_1 = os.path.join("../../data/",config["FASTQS"]["read1"] + ".trimmed.fastq"),
        trim_2 = os.path.join("../../data/",config["FASTQS"]["read2"] + ".trimmed.fastq")
    shell:
        "cutadapt --minimum-length=5 --output {output[0]} "
        "--paired-output {output[1]} -A XXX -q 30 {input}"

if not os.path.exists(os.path.join(config["GENOME_DIR"], "STAR_INDEX")):
    os.makedirs(os.path.join(config["GENOME_DIR"], "STAR_INDEX"))

rule star_refIndex:
    input:
        os.path.join(config["GENOME_DIR"], config["GENOME_FASTA"]),
        os.path.join(config["GENOME_DIR"], config["GENOME_GTF"]),
        os.path.join(config["GENOME_DIR"],"STAR_INDEX/")
    output:
        os.path.join(config["GENOME_DIR"], "STAR_INDEX", "Genome"),
        os.path.join(config["GENOME_DIR"], "STAR_INDEX", "SA"),
        os.path.join(config["GENOME_DIR"], "STAR_INDEX", "SAindex")
    shell:
        "STAR --runThreadN 16 --runMode genomeGenerate "
        "--genomeDir {input[2]} "
        "--genomeFastaFiles {input[0]} "
        "--sjdbGTFfile {input[1]}"
