"""
__authors__: Kevin Chau, Weixiu Dong, Hanqing Zhao
__date__: 2019-03-05
__description__: RNA-Seq alignment and quantification
"""

import os
configfile: "../../config.yaml"
# Sub-Workflows

subworkflow align_and_quant:
    workdir: "workflows/align_and_quant"
    snakefile: "workflows/align_and_quant/Snakefile"  # Not required, but helpful
    configfile: "../../config.yaml"

if not os.path.exists("../../data/STAR_ALIGN"):
    os.makedirs("../../data/STAR_ALIGN")

rule run_star:
    input:
        os.path.join(config["GENOME_DIR"], "STAR_INDEX", "Genome"),
        os.path.join(config["GENOME_DIR"], "STAR_INDEX", "SA"),
        os.path.join(config["GENOME_DIR"], "STAR_INDEX", "SAindex"),
        os.path.join("../../data/",config["FASTQS"]["read1"] + ".trimmed.fastq"),
        os.path.join("../../data/",config["FASTQS"]["read2"] + ".trimmed.fastq"),
        os.path.join(config["GENOME_DIR"], "STAR_INDEX/")
    output:
        os.path.join("../../data/STAR_ALIGN", config["FASTQS"]["prefix"] + ".bam")
    shell:
        "STAR --genomeDir {input[5]}"
        " --runThreadN 12"
        " --readFilesIn {input[3]} {input[4]}"
        " --outFilterType BySJout"
        " --outFilterMultimapNmax 20"
        " --alignSJoverhangMin 8"
        " --alignSJDBoverhangMin 1"
        " --outFilterMismatchNmax 999"
        " --alignIntronMin 20 --alignIntronMax 1000000"
        " --alignMatesGapMax 1000000 --outSAMunmapped Within"
        " --outSAMattributes NH HI AS NM MD"
        " --outFilterMismatchNoverReadLmax 0.04 --sjdbScore 1"
        " --outSAMtype BAM SortedByCoordinate --quantMode TranscriptomeSAM"
        " --outFileNamePrefix ../../data/STAR_ALIGN/"+config["FASTQS"]["prefix"]+".STAR"
